File: ios/Solace/ViewModels/EmotionViewModel.swift
import Foundation
import FirebaseFirestore

@MainActor
final class EmotionViewModel: ObservableObject {

    @Published var logs: [MoodLog] = []
    @Published var isLoading: Bool = true
    @Published var isSaving: Bool = false
    @Published var savedToday: Bool = false
    @Published var selectedEmoji: String = "😐"
    @Published var selectedMoodKey: MoodColorKey = .calmRelaxed
    @Published var note: String = ""
    @Published var error: String? = nil

    private var currentUid: String = ""
    private var listener: ListenerRegistration?
    private let repo = MoodRepository.shared

    func initialize(uid: String) {
        guard uid != currentUid else { return }
        currentUid = uid
        listener?.remove()
        listener = repo.observeLogs(uid: uid) { [weak self] logs in
            guard let self else { return }
            self.logs = logs
            self.isLoading = false
        }
    }

    func logMood() {
        isSaving = true
        Task {
            do {
                try await repo.addLog(
                    uid: currentUid,
                    emoji: selectedEmoji,
                    moodColorKey: selectedMoodKey.rawValue,
                    note: note.trimmingCharacters(in: .whitespacesAndNewlines)
                )
                note = ""
                savedToday = true
            } catch {
                self.error = error.localizedDescription
            }
            isSaving = false
        }
    }

    func dismissConfirmation() { savedToday = false }

    deinit { listener?.remove() }
}

File: android/app/src/main/java/com/solace/ui/emotion/EmotionViewModel.kt
package com.solace.ui.emotion

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.solace.data.emotion.MoodLog
import com.solace.data.emotion.MoodRepository
import com.solace.ui.config.MoodColorKey
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import javax.inject.Inject

data class EmotionUiState(
    val logs: List<MoodLog> = emptyList(),
    val isLoading: Boolean = true,
    val isSaving: Boolean = false,
    val savedToday: Boolean = false,  // shows confirmation after logging
    val selectedEmoji: String = "😐",
    val selectedMoodKey: MoodColorKey = MoodColorKey.CALM_RELAXED,
    val note: String = "",
    val error: String? = null,
)

@HiltViewModel
class EmotionViewModel @Inject constructor(
    private val repo: MoodRepository,
) : ViewModel() {

    private val _state = MutableStateFlow(EmotionUiState())
    val state: StateFlow<EmotionUiState> = _state.asStateFlow()

    private var currentUid = ""

    fun init(uid: String) {
        if (currentUid == uid) return
        currentUid = uid
        viewModelScope.launch {
            repo.getLogs(uid).collect { logs ->
                _state.update { it.copy(logs = logs, isLoading = false) }
            }
        }
    }

    fun selectEmoji(emoji: String) = _state.update { it.copy(selectedEmoji = emoji) }
    fun selectMood(key: MoodColorKey) = _state.update { it.copy(selectedMoodKey = key) }
    fun setNote(note: String) = _state.update { it.copy(note = note) }
    fun clearError() = _state.update { it.copy(error = null) }
    fun dismissConfirmation() = _state.update { it.copy(savedToday = false) }

    fun logMood() {
        val s = _state.value
        viewModelScope.launch {
            _state.update { it.copy(isSaving = true) }
            repo.addLog(
                uid = currentUid,
                emoji = s.selectedEmoji,
                moodColorKey = s.selectedMoodKey.name,
                note = s.note.trim(),
            ).onSuccess {
                _state.update { it.copy(isSaving = false, savedToday = true, note = "") }
            }.onFailure { e ->
                _state.update { it.copy(isSaving = false, error = e.message) }
            }
        }
    }
}
